### PR TITLE
Lets 'ipfs add' include top-level hidden files

### PR DESCRIPTION
**Pull request by noffle**
Fixes ipfs/go-ipfs/#2145. The --hidden switch (still) only affects recursive adding.

License: MIT
Signed-off-by: Stephen Whitmore <noffle@ipfs.io>

### CODE REVIEW

**Comment by jbenet** [comment]
maybe hidden files should act on non-recursive adds too, unless `-f`?

```sh
> mkdir /tmp/foo
> cd /tmp/foo
> git init
> touch bar
> echo "bar" >.gitignore
> git add bar
The following paths are ignored by one of your .gitignore files:
foo
Use -f if you really want to add them.
```

File: core/commands/add.go [hunk 1]
```diff
@@ -147,8 +148,20 @@ remains to be implemented.
 		fileAdder.Silent = silent
 
 		addAllAndPin := func(f files.File) error {
-			if err := fileAdder.AddFile(f); err != nil {
-				return err
+			// Iterate over each top-level file and add individually. Otherwise the
+			// single files.File f is treated as a directory, affecting hidden file
+			// semantics.
```
Comment (jbenet) [inline core/commands/add.go]: good catch!

File: test/sharness/t0042-add-skip.sh [hunk 2]
```diff
@@ -48,6 +48,17 @@ test_add_skip() {
 		test_cmp expected actual
 	'
 
+	test_expect_success "'ipfs add' includes hidden files at the top-level even without --hidden" '
+    mkdir -p mountdir/dotfiles &&
+    echo "set nocompatible" > mountdir/dotfiles/.vimrc
+		cat >expected <<-\EOF &&
+added QmT4uMRDCN7EMpFeqwvKkboszbqeW1kWVGrBxBuCGqZcQc .vimrc
+		EOF
+		ipfs add mountdir/dotfiles/.vimrc >actual
```
Comment (jbenet) [inline test/sharness/t0042-add-skip.sh]: yeah i think this should be `ipfs add -f mountdir/dotfiles/.vimrc` -- thoughts?

**Comment by jbenet** [comment]
Sorry, i'm being silly. this is for _hidden_ files, not for _ignored_ files.

File: test/sharness/t0042-add-skip.sh [hunk 3]
```diff
@@ -48,6 +48,17 @@ test_add_skip() {
 		test_cmp expected actual
 	'
 
+	test_expect_success "'ipfs add' includes hidden files at the top-level even without --hidden" '
```
Comment (jbenet) [inline test/sharness/t0042-add-skip.sh]: maybe say `'ipfs add' includes hidden files given explicitly even without --hidden` -- top-level was a bit confusing given the example below (`mountdir/dotfiles/.vimrc`), took me a second to realize it meant "top level of the commands file tree"

**Comment by jbenet** [comment]
one way to do the "explicit/direct" thing is to have the commands lib label the files that are added directly vs those added recursively, then the adder can check something like:

 ```go
if f.Explicit() || adder.Hidden() { ... }
```

**Comment by jbenet** [comment]
@noffle comments above, otherwise LGTM. +1 on comments, thank you.

**Comment by noffle** [comment]
> one way to do the "explicit/direct" thing is to have the commands lib label the files that are added directly vs those added recursively

Happy to have you sell me on it, but I *really* don't like the idea of the exposing the semantics of how command line processing works to the FileAdder -- I think this sort of "hidden unless an 'explicit' file" notion should live on the outside.

File: test/sharness/t0042-add-skip.sh (hunk 3 above)
Comment (noffle) [inline test/sharness/t0042-add-skip.sh]: You're 100% right -- that's pretty confusing. I'll steal yours. :)

**Comment by jbenet** [comment]
Oh i understand now-- i was missing that _explicitly added single files will **always** be in top level_, obvious in retrospect.

**Comment by jbenet** [comment]
ok LGTM! -- can you squash 01bb593 into 15554b1 (or squash the whole thing?) we have a requirement in this repo that all commits should pass the tests so that we can `git-bisect` easily.

**Comment by jbenet** [comment]
(last thing and good to merge i think!

**Comment by whyrusleeping** [comment]
cool, this LGTM

**Comment by noffle** [comment]
> we have a requirement in this repo that all commits should pass the tests so that we can git-bisect easily.

That is an excellent requirement. Can do! :confetti_ball:

**Comment by whyrusleeping** [comment]
cool cool cool :ship:

**Pull request merged**